### PR TITLE
Use paths in codebase

### PR DESCRIPTION
**Pull request by pombredanne**
This PR merges the latest skeleton and implements new internals for resource Codebase and Resource.
The key change is dropping using numeric resource ids and using a simpler map of path->Resource instead.
The not-yet-implemented part is focusing a codebase on the subset the new `Codebase(paths)` argument. 
This is still a work in progress because of this. But early feedback is welcomed and needed.

### CODE REVIEW

**Review by JonoYang — COMMENTED** [review]
I'm running into errors when using the `VirtualCodebase`. I am passing in two JSON scans into `VirtualCodebase` to test the functionality we have where we can create a `VirtualCodebase` from multiple scans. I am able to create the `VirtualCodebase` but using the `walk()` method causes an exception to be raised:

```
Traceback (most recent call last):
  File "/home/jono/nexb/src/commoncode/src/commoncode/resource.py", line 1228, in children
    return sorted(children, key=_sorter)
  File "/home/jono/nexb/src/commoncode/src/commoncode/resource.py", line 1220, in <lambda>
    _sorter = lambda r: (r.has_children(), r.name.lower(), r.name)
AttributeError: 'NoneType' object has no attribute 'has_children'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jono/nexb/src/commoncode/src/commoncode/resource.py", line 867, in walk
    for res in root.walk(self, topdown=topdown, ignored=ignored):
  File "/home/jono/nexb/src/commoncode/src/commoncode/resource.py", line 1196, in walk
    ignored=ignored,
  File "/home/jono/nexb/src/commoncode/src/commoncode/resource.py", line 1187, in walk
    for child in self.children(codebase):
  File "/home/jono/nexb/src/commoncode/src/commoncode/resource.py", line 1230, in children
    raise Exception(f'Cannot sort children: {children!r}:\n{children_paths!r}') from e
Exception: Cannot sort children: [None, None]:
['codebase/package', 'codebase/django-audit-tools-0.4.0']
```

I took a look at the paths of the `VirtualCodebase` I created by using the `resources_by_path`  attribute and I saw that the paths root Resources of the two scans I used do not start with `virtual_root`, whereas all of the other Resources in the `VirtualCodebase` do. I think that the exception is occurring because the Resources doesn't have `virtual_root` in its path and it cannot find any other children because of the difference in the path prefix.

**Comment by pombredanne** [comment]
> I'm running into errors when using the `VirtualCodebase`. I am passing in two JSON scans into `VirtualCodebase` to test the functionality we have where we can create a `VirtualCodebase` from multiple scans. I am able to create the `VirtualCodebase` but using the `walk()` method causes an exception to be raised:

@JonoYang the latest push is fixing this. And several other issues. Thanks!

File: src/commoncode/resource.py [hunk 1]
```diff
@@ -242,39 +228,55 @@ class Codebase(object):
...
 
-        `max_in_memory` is the maximum number of Resource instances to keep in
+        ``max_in_memory`` is the maximum number of Resource instances to keep in
         memory. Beyond this number, Resource are saved on disk instead. -1 means
         no memory is used and 0 means unlimited memory is used.
 
-        `max_depth` is the maximum depth of subdirectories to descend below and
+        ``max_depth`` is the maximum depth of subdirectories to descend below and
         including `location`.
         """
```
Comment (AyanSinhaMahapatra) [inline src/commoncode/resource.py:281]: We could add a line about `paths` in the docstring.

File: src/commoncode/resource.py [hunk 2]
```diff
@@ -285,36 +287,44 @@ def __init__(self, location,
...
+        """
+        paths = (clean_path(p) for p in (paths or []) if p)
+        # we sort by path segments (e.g. essentially a topo sort)
+        _sorter = lambda p: p.split('/')
+        return sorted(paths, key=_sorter)
+
     def _setup_essentials(self, temp_dir=temp_dir, max_in_memory=10000):
         """
-        Set the remaining Codebase attributes
+        Set the remaining Code=base attributes
```
Comment (AyanSinhaMahapatra) [inline src/commoncode/resource.py:327]:
```suggestion
        Set the remaining Codebase attributes
```

File: src/commoncode/resource.py [hunk 3]
```diff
@@ -1181,40 +1292,78 @@ def walk(self, codebase, topdown=True, ignored=ignore_nothing):
...
+        children = [get_child(codebase, name) for name in children_names]
+
         _sorter = lambda r: (r.has_children(), r.name.lower(), r.name)
-        get_resource = codebase.get_resource
-        return sorted((get_resource(rid) for rid in self.children_rids), key=_sorter)
+        return sorted((c for c in children if c), key=_sorter)
 
     def has_parent(self):
         """
         Return True is this Resource has children.
```
Comment (AyanSinhaMahapatra) [inline src/commoncode/resource.py:1351]:
```suggestion
        Return True is this Resource has a parent.
```

**Review by AyanSinhaMahapatra — APPROVED** [review]
Looks great! Just a few nits.

**Review by JonoYang — APPROVED** [review]
I think it looks okay. I like that you set the __iter__ dunder on Codebase to yield its resources. 

I feel that the solution we have for creating a codebase from multiple scans is a bit kludgy, but reasonable. Treating each scan as an individual codebase is easier than attempting to always merge scan directory structure/data together. I had not realized the implication of getting scancode to merge two directories named `codebase` from two different scans. 

On a whim, I've updated commoncode in scancode and all the references to `rid` in `scancode/cli.py` will need to be updated.

**Pull request merged**